### ISSUE TITLE
[Feature request]: Add new syntax for jsx classes

### ISSUE THREAD

**Issue by MaxmaxmaximusGitHub** [reactjs/rfcs#155]
I remind you that JSX itself is a SUGAR on top of javascript ;)
---

💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡
💡 **We need more sugar! New syntax for jsx classes!** (and isolated styles)
💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡💡

- Start Date: 2020-04-13

# Summary

new JSX class syntax:
1) attr start with dot is class name (intentionally similar to css, yes)
2) if this attr have expression value, then if exp is true, then class exists, otherwise not exist
3) **of course, these classes have an ISOLATED SCOPE, for the ISOLATION of styles, I came up with this syntax.**

# Basic example

this:

```javascript
<div .block .chat .active={ state.active }></div>
```

compile to:

```javascript
<div className={ `block chat${ state.active ? " active" : "" }` } ></div>
```

by the way, no one forbids you to add dynamic class names, and still use your favorite css modules

this:

```jsx
<div .[css.button] .[css.active]={ user.active } ></div>
```

compile to:

```jsx
<div className={ `${css.button}${ user.active ? " " + css.active : "" }` }></div>
```

# Motivation

**Attributes names with started a dot symbol is prohibited, so dot is idle idle. dot can be used for syntactic sugar for routine 90% of the operations of assigning classes to elements.**

people will be able to write a main class describing what kind of entity it is. and add dynamic modifier classes, which reflect the component state.

it’s just syntactic sugar. but it will save 90% of the time for developers. the programmer simply type on a keyboard the dot symbol, and the IDE will automatically autocomplete them for the available classes.

in brevity and elegance, this syntax is even superior to html ;)

our beloved jsx is high time to develop, because competitors (angular and vue) do not stand still ;)

# Detailed design

you just look how cool:
```jsx
<ul .users>
  {this.props.users.map( user =>
      <li .user .banned={ user.isBanned } .admin={ user.isAdmin }>
        User { user.name }
      </li>
  )}
</ul>
```

and imagine how it will be cool with syntax highlighting in the IDE and if you click by mouse wheel you will immediately be transferred to the desired selector in css file.

what it looks like now:

```jsx
<ul className="users">
  {this.props.users.map( user =>
      <li className={ `user${ user.isBanned ? " banned" : "" }${ user.isAdmin ? " admin" : "" }` }>
        User { user.name }
      </li>
  )}
</ul>
```

feel the difference?
===

this:

```jsx
<div .box .active={active} onClick={toggleActive}></div>
```

compile to this:

```jsx
<div className={ `box${ active ? " active" : "" }` } onClick={toggleActive}></div>
```

just compare
---

```jsx
export default function TabTitle({active, onClick, children}) {

  return <div className={ classes('tab-title', {
    __active: active
  }) } onClick={ onClick }>

    { children }

    <style jsx>{ style }</style>
  </div>
}
```

and my version:
```jsx
export default function TabTitle({active, onClick, children}) {

  return <div .tab-title .__active={active} onClick={ onClick }>

    { children }

    <style jsx>{ style }</style>
  </div>
}
```

is a syntax hell :
---
![image](https://user-images.githubusercontent.com/60403187/80319125-5af74000-8817-11ea-8a8e-da62b82bde9c.png)

We can use new syntax and old syntax at the same time:
---

this:

```jsx
function classes(classesArr){ classesArr.join(' ') }

<div .box .active={active} className={ cssModule.class + classes(['one', 'two']) }>
```

compile to:

```jsx
function classes(classesArr){ classesArr.join(' ') }

return <div className={[cssModule.class + classes(['one', 'two']), `box${active ? ' active' : '' }`].join(' ') }>
```

# Drawbacks

Why should we *not* do this? Please consider:

There are no drawbacks, since the dot symbol was previously prohibited, backward compatibility is present.

The disadvantage is that jsx with the new syntax will cause a syntax error in compilers that do not support the new function. therefore, we must write babel-plugin to translate the new code into the "old" jsx.

# Alternatives

You can write the babel plugin to support the new syntax, but this will not motivate the creators of the IDE to add new syntax. The popularity of React and Facebook and the headline "React 17.0 - New Syntax for Classes." will motivate support  this feature.

# Adoption strategy

React competitors use much more convenient syntax than jsx. Since it uses attributes starting with special characters. jsx does not use such attributes, and simply prohibits them. People are tormented by class definitions in jsx and rejoice in Angular and Vue. People will be happy with the new “meaningful” syntax, which will be much more organic than native html. Since in css classes start with a dot, in jsx they will also start with a dot. And also the most common operation with classes is adding or removing a class, depending on the state. We add and make it native. There is not a single developer in the world who says "bliiin, well, support for convenient classes has appeared (" and there are 99% of people who say "FINALLY" =)

# How we teach this

This innovation is sugar, and does not break or change existing behavior. Therefore, we can simply ADD a new section to the "dot classes" documentation. Those who like the new syntax will use it, those who like the old will use the old. Anyone who likes both syntaxes can combine them. className is just a string, so at the jsx compilation stage we can quietly unify both methods of class declaration.

# Unresolved questions

Since we are doing such a thing, then maybe we should immediately think about isolated styles for the component? Since we are making syntax changes, it may be that we think over the isolation of styles in advance? After all, if we implement this without thinking it over, then we will not be able to change it on the future. I suggest thinking about how to natively (from the point of view of jsx) and syntactically solve the problem of encapsulating classes, and how to make it flexible enough so that other encapsulation libraries can USE this syntax.

by the way, we can also add dynamic attribute names
===

this:

```jsx
let prop = "ololo"
<div [prop]="value"></div>
```

compile to:

```javascript
let prop = "ololo"
React.createElement('div', { [prop]: "value"} )
```

![image](https://user-images.githubusercontent.com/60403187/80296668-ad881c00-8785-11ea-85d7-79abc6996f46.png)

**Comment by MaxmaxmaximusGitHub** [comment]
I think style isolation is the work of the css and html standard and not ours. we, as a temporary solution, can do this: babel может найти все dot classes in code, and add to them namespace/ like this:


`button.css:`
```css
.button {
  color: red;
}
```

`button.js:`
```javascript
import './button.css'

export default function Button(){ 
  return <button .button>{ this.props.childiren }</button>
}
```


compile to:


`button.css:`
```css
.button__namespace__ {
  color: red;
}
```

`button.js:`
```javascript
import './button.css'

export default function Button(){ 
  return <button .button__namespace__>{ this.props.childiren }</button>
}
```

**Comment by MaxmaxmaximusGitHub** [comment]
as a workaround NOW until dot classes are implemented. I recommend this solution:

this:
---
`button.css:`
```css
.button {
  color: red;
}
```

`button.js:`
```javascript
import './button.css'

export default function Button(){ 
  return <button className="button">{ this.props.childiren }</button>
}
```


compile to:
---

`button.css:`
```css
.button[__namespace__] {
  color: red;
}
```

`button.js:`
```javascript
import './button.css'

export default function Button(){ 
  return <button [__namespace__]="" className="button">{ this.props.childiren }</button>
}
```

**Comment by foray1010** [comment]
I downvoted this because:
1. It doesn't stick to HTML syntax, it is hard to understand for new comer
2. It is impossible to type it in flow or typescript as it is a dynamic prop name

**Comment by MaxmaxmaximusGitHub** [comment]
@foray1010 

> It doesn't stick to HTML syntax, it is hard to understand for new comer
1) it follows the html syntax, since ANY characters in attribute names are allowed in html.
2) className also does not adhere to the html syntax since in html you need to write a class and not className
3) jsx itself does not adhere to the html syntax since in html the value of attributs should be framed in quotes, and jsx allows the expression `attr = {2 + 3}`
4) The new syntax will not cause confusion among the beginner, but will increase the understanding of beginners, therefore you are wrong.

Therefore, your arguments are not valid, therefore, like.

> It is impossible to type it in flow or typescript as it is a dynamic prop name

purpose, sentences, to make class names static, not dynamically, therefore, you are completely diametrically misunderstood this sentence, so set like.

in other words, everything you said is wrong.
your assessment is based not on the flaws of my proposal, but on your misunderstanding. =)

**Comment by foray1010** [comment]
Did u use typed language before? we need to [predefined attribute types](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L1743-L1809) before use

**Comment by mAAdhaTTah** [comment]
@foray1010 This new JSX syntax would transpile down to `className`, so the TS/Flow issue wouldn't be an issue.

That said, JSX isn't DOM-specfic, so I'm not sure how acceptable new syntax that targets `className` specifically would be here.

**Comment by j-f1** [comment]
I dislike this because it favors one means of styling (CSS using global class names) over all others (CSS modules, inline styles, CSS-in-JS, React Native, etc.) for no clear reason, and blocks JSX from ever using this syntax for something else in the future.

**Comment by MaxmaxmaximusGitHub** [comment]
@j-f1 no, this syntax is CREATED for isolated and local class names, you completely did not understand the idea.

**Comment by MaxmaxmaximusGitHub** [comment]
@foray1010 
> Did u use typed language before? we need to predefined attribute types before use

dot classes are not html attribute names. you just did not understand the idea

**Comment by MaxmaxmaximusGitHub** [comment]
@mAAdhaTTah 

> That said, JSX isn't DOM-specfic, so I'm not sure how acceptable new syntax that targets className specifically would be here. 

now attributes starting with a period are generally prohibited. no platform uses them. There are no conflicts with other platforms, just a bonus for the web platform. benefit is - there is no harm.

**Comment by mAAdhaTTah** [comment]
Unless the feature is broadly useful, I don't see this being accepted as part of JSX. You implicitly recognize there's no use for it in React Native, and there is also no clear use for it in other rendering targets like canvas, webgl, & the terminal, all of which have React renderers. New syntax that's only useful in one case doesn't really clear the bar for adding it.

**Comment by dantman** [comment]
> @j-f1 no, this syntax is CREATED for isolated and local class names, you completely did not understand the idea. I, using this syntax in my framework, work with a bang https://www.npmjs.com/package/ui-js

Your first post gives this example:

> ```js
> <div .block .chat .active={ this.state.active }></div>
> ```
> 
> 
> its equivalent for
> ```js
> <div className={ `block chat${ this.state.active ? " active" : "" }` } ></div>
> ```

How are these global class names "isolated and local class names"?

**Comment by MaxmaxmaximusGitHub** [comment]
@mAAdhaTTah 

> Unless the feature is broadly useful

it is very useful, since now it is inconvenient to make dynamic classes based on variables. and this function will make it convenient. this function will save 40% of the time during layout. Is saving 40% of the time in layout not a useful feature?

**Comment by MaxmaxmaximusGitHub** [comment]
> How are these global class names "isolated and local class names"?

These are not global class names, these are local class names. the programmer should not be involved in the manual addition of namespace, so you mistakenly assumed that these are global class names.

Next, I wrote:
1) Now that we have a static definition of class names, we can ANALYZE them at the stage of compiling the code, and we can add namespaces. In other words, the syntax proposed by me makes the work of transpilers that add namespaces and isolation systems to classes easier. Since I came up with a STANDARD for class descriptions.

Then I wrote:

1) Since we made the class description declarative and created a special syntax for it, we can also come up with a class isolation system and build it into the react, or leave this to the Babel type transporters.

You just did not understand my PHENOMINAL idea, like everyone who put dislike. Their programming level is fundamentally low compared to mine. Therefore, they are not able to realize such a revolutionary breakthrough and useful approach. They are still young, learn later.

**Comment by phryneas** [comment]
Are you a troll or really just extremly rude and patronizing?

**Comment by MaxmaxmaximusGitHub** [comment]
@phryneas I just write a translator through Google, apparently he translates everything verbatim

**Comment by MaxmaxmaximusGitHub** [comment]
@mAAdhaTTah 

> New syntax that's only useful in one case doesn't really clear the bar for adding it.

Now the dot symbol is not used at all, what’s wrong with adding it as sugar for the className property for the web platform? no, literally. give a hypothetical situation when this could harm the jsx language? give an example of a situation where we would be sad that we had standardized such syntax, and then regretted it?

**Comment by phryneas** [comment]
> You just did not understand my PHENOMINAL idea, like everyone who put dislike. Their programming level is fundamentally low compared to mine. 

This has nothing to do with translation. Putting your ideas above others is foolish at best, but actually judging and devaluing the "programming skills" of others because they do not agree to you is rude in every spoken language on this planet.

Liking your own posts does not make it better.

But let's get back to discussing this topic.

Essentially your suggestion is to use [css modules](https://github.com/css-modules/css-modules) (which are already a thing) in an implicit combination between the named import `style`  with a new jsx syntax in the same file.

So instead of writing 
```js
<li classname={clsx(style.user, { [style.banned]: user.isBanned, [style.admin]: user.isAdmin }>
```
one would write
```js
<li .user .banned={ user.isBanned } .admin={ user.isAdmin }>
```

This is arguably a bit shorter, but imposes quite a lot of problems:
* it introduces an implicit relation between the component and the import `style` that is not visible to the developer on first sight
* it will break in code where the variable `style` is already defined
* this new syntax could be transpiled away with babel, *but* it would break all existing tooling:
eslint, prettier, typescript and a lot other developer-level would all read this code before it ever passed babel, so they would need to be modified to work with it
* this new syntax would now be supported by a lot of tooling, but at the same moment be incompatible with other frameworks that also use jsx, but not the `classname` property. React is not the only framework that uses jsx and as far as I know, `classname` is not standardized anywhere.
* this would somehow "officially embrace" css modules over other valid approaches like 
  * global css file
  * css frameworks like [tailwind css](https://tailwindcss.com/)
  * css-in-js frameworks like [styled-components](https://styled-components.com/), [emotion](https://github.com/emotion-js/emotion), or other [jss](https://cssinjs.org/?v=v10.1.1)-based solutions like material-ui's [useStyle](https://material-ui.com/styles/basics/) hook 
Each of these solutions can, depending on the situation, be far better or worse than each other of these solutions. Embracing one over all others would be a disservice to the community.


And all that to save a few characters - which is something that, if you are using a good component library and writing clean code should not be much of a concern anyways.

So you see, there are quite some good reasons people might be sceptical about your suggestion, and someone not agreeing to you might not be an indication of them being stupid, but maybe just an indication that they have other concerns than you.

Personally, I would definitely not call your idea "phenomenal". But that's just my opinion.

**Comment by MaxmaxmaximusGitHub** [comment]
@phryneas 
> Essentially your suggestion is to use css modules

No, if you use css modules use them as before. Their syntax is extremely ugly, which made me develop a new syntax.

Those who don't like the ugly syntax of css modules can use my new syntax, getting concise, concise, and ISOLATED styles.

Moreover, both approaches can be used simultaneously. In other words, you can use as css modules using className, and at the same time use the new beautiful syntax.

I developed my ingenious syntax without breaking backward compatibility in order to maintain backward compatibility. After all, I am a professional programmer who first of all takes into account the needs of the ecosystem.

Example:
---

this:

```jsx
function classes(classesArr){ classesArr.join(' ') }

<div .box .active={active} className={ cssModule.class + classes(['one', 'two']) }>
```

compile to this:

```jsx
function classes(classesArr){ classesArr.join(' ') }

return <div className={[cssModule.class + classes(['one', 'two']), `box${active ? ' active' : '' }`].join(' ') }>
```

**Comment by phryneas** [comment]
> there is not a single reason, and all the reasons that you called are wrong, and they are based only on your misconceptions

So neither eslint, nor prettier or typescript will have to adopt their parser and release a new version to support this? Babel must be magic, even if it isn't being used.

**Comment by MaxmaxmaximusGitHub** [comment]
> phryneas

typescript and babel NEVER do new syntax development, why would they do it if Facebook supports and standardizes it? [sarcasm]

**Comment by phryneas** [comment]
> I remind you that you still did not mention a single minus in my idea

I mentioned many, you answered condescendingly on one and ignored the rest. 

I'm out. 

"RFC" stands for "Request For Comment". 
Obviously you are ignoring the biggest part of the comments while still somehow managing to be rude, so there's no point in trying to have any meaningful discussion here.

**Comment by MaxmaxmaximusGitHub** [comment]
@phryneas  

> it introduces an implicit relation between the component and the import style that is not visible to the developer on first sight

does not introduce

> it will break in code where the variable style is already defined

not break 

> this new syntax could be transpiled away with babel, but it would break all existing tooling:
eslint, prettier, typescript and a lot other developer-level tools would all read this code before it ever passed babel, so they would need to be modified to work with it

for the same reason, you can not add new syntax to all languages at all. the argument is not consistent and absurd.

> his new syntax would now be supported by a lot of tooling, but at the same moment be incompatible with other frameworks that also use jsx, but not the classname property. React is not the only framework that uses jsx and as far as I know, classname is not standardized anywhere.

if your platform does not support the className property then why are you using it? (even as sugar "dot classes")


> this would somehow "officially embrace" css modules over other valid approaches like

"dot classes" is just sugar on top of className, they are not connected in any way with any platform or framework and do not do anything you could not write with your hands. I already recommended that you familiarize yourself with the concept of syntactic sugar, I highly recommend reading the link from Wikipedia.

**Comment by juzhiyuan** [comment]
just unsubscribed...

**Comment by MaxmaxmaximusGitHub** [comment]
@MaxmaxmaximusGitHub 

here smart and serious people decide how to develop the jsx syntax, I think you should just unsubscribe and not clutter the discussion

**Comment by juzhiyuan** [comment]
how about id selector and data attribute? according to this proposal, we may use the following syntax?

```jsx
<div #awsome-id data-a="xxx" />
```

HTML has too many attributes I think...

**Comment by MaxmaxmaximusGitHub** [comment]
@juzhiyuan


> how about id selector and data attribute? according to this proposal, we may use the following syntax in the future?

The idea is good, it is already implemented in my framework. But in the case of jsx this would not be appropriate. Indeed, there are no such problems with writing id = "myID" as with writing classes. Therefore, there is no particular reason to introduce a new syntax, and to occupy a whole sign under it.

**Comment by MaxmaxmaximusGitHub** [comment]
@juzhiyuan


so far I have heard only one valid argument against such syntax. the fact that className only makes sense for the web, and making special sugar for the web is not cool.

but I parried it like this: now attributes from the point are generally prohibited, the point is idle, and at least it will bring joy to the web platform =)

**Comment by MaxmaxmaximusGitHub** [comment]
@phryneas  

and by the way, no one forbids you to add dynamic class names.

```jsx
<div .[css.button] .[css.active]={ user.active } ></div>
```

**Comment by dancerphil** [comment]
We are use jsx based on __Components__ more than __html__.

And I think `className` is only a tiny bit one of all of the props. Not to mention that it can only be used in html-based components.

It is not compatible with the overall idea of react, and more likely another anti-pattern.

**Comment by MaxmaxmaximusGitHub** [comment]
@dancerphil 

> We are use jsx based on Components more than html.
then why do you need to use className? do not use

> It is not compatible with the overall idea of react

The general idea of a react in declarative reactivity is when the function is executed again and does not have side effects. what does the syntactic sugar for the className property have to do with it, and what idea does syntactic sugar for the className property contradict?

> and more likely another anti-pattern.

What harm will ecosystem sugar have for the className property?

**Comment by dancerphil** [comment]
> **Why from this we should not add syntactic sugar for popular use in one of the platforms?**

We should add it __if and only if__ html adds it as a syntactic sugar. Otherwise we shouldn't.

**Comment by MaxmaxmaximusGitHub** [comment]
> We should add it if and only if html adds it as a syntactic sugar. Otherwise we shouldn't.
Why?

**Comment by MaxmaxmaximusGitHub** [comment]
so, here’s something else I’ve come up with, I hope you enjoy it more https://github.com/reactjs/rfcs/issues/161

**Comment by MaxmaxmaximusGitHub** [comment]
![image](https://user-images.githubusercontent.com/60403187/80306389-aedc3780-87cb-11ea-8b4b-408a1f55f4ee.png)

list of the smartest and most educated people on the planet <3

**Comment by MaxmaxmaximusGitHub** [comment]
just compare
---

```jsx
export default function TabTitle({active, onClick, children}) {

  return <div className={ classes('tab-title', {
    __active: active
  }) } onClick={ onClick }>

    { children }

    <style jsx>{ style }</style>
  </div>
}
```

and my version:
```jsx
export default function TabTitle({active, onClick, children}) {

  return <div .tab-title .__active={active} onClick={ onClick }>

    { children }

    <style jsx>{ style }</style>
  </div>
}
```

**Comment by gaearon** [comment]
Hi @MaxmaxmaximusGitHub,

Thank you for your suggestion.
Generally saying, RFCs should be sent as pull requests, but I’ll reply here.

The scope of this RFC is much larger than just React. At the very least, it involves changes to JSX which we discuss [here](https://github.com/facebook/jsx/issues). As @phryneas noted, any changes to JSX are a lot of work for everyone involved to push through because they need to be coordinated with many partner projects, including Babel, ESLint, Prettier, TypeScript, Flow, IDEs and syntax highlighters, and so on. This doesn’t mean it’s insurmountable, but just that the bar for adding something to the JSX grammar is very high.

Any additions to JSX should be in harmony with other things we’ve been considering changing there. For example, syntax like `.block .chat .active={state.active}` presumably matches how currently `foo bar` become `foo={true} bar={true}`. However, with https://github.com/facebook/jsx/issues/23 that would not be the case, where they would become `foo={foo} bar={bar}`. It is not clear whether that change is actually worth it, but it is far more commonly requested so this contradiction is something to consider. We can set that aside though.

I think this proposal can be divided in two parts:

1. A way to conditionally apply classes without wrangling strings
2. A first-class "CSS class" marker to be recognized by tooling

If (1) is your primary problem, you can already do this today with helpers like classnames:

```jsx
import c from 'classnames';

<ul className="users">
  {this.props.users.map(user =>
    <li className={c('user', { banned: user.isBanned, admin: user.isAdmin )}>
      User {user.name}
    </li>
  )}
</ul>

<div className={c('box', { active })} onClick={toggleActive} />

<div className={c(css.button, user.active && css.active)} />
```

This seems equivalent in terms of how expressive it is.

If it's typing `className` that gets you, you could have a helper that you can spread:

```jsx
<div {...c('box', { active })}>
```

This is somewhat cryptic, but your proposal also favors brevity.

But maybe this isn’t enough and you’re really after (2) — a first-class marker for CSS class names in syntax. I think there are a few problems with this.

Even in your example, I found `<div .user>` and `<div user>` somewhat difficult to tell apart. Sure, this can be "fixed" by different syntax highlighting, but this is so easy to miss in text that I don’t think it’s desirable. In the examples you’ve provided it’s less of a problem because those are clearly not `<div>` attributes, but it gets easier to get confused when you’re dealing with custom components:

```jsx
<User .user user={user} .active />
```

Another problem with this is the clash resolution.

```jsx
const props = {
  className: 'foo'
}
<div .bar {...props} />
```

We can't know statically what `props` contains, so we need to figure out what to do here. Do we ignore spread `className` in props and force you to pull it out statically? Do we forbid `.bar` syntax together with spread? Do we join classes?

Your proposal seems to prefer joining classes. We actually did that as a part of `transferProps()` helper in very early versions of React (around React 0.12 era). This has created a bunch of problems because in practice people often wanted to override them instead, and the joining was annoying. Spread syntax adds a layer of complexity to that already thorny situation.

If we pass `<Foo .a={true} .b={false} />`, and then it renders `<Bar {...props} .a={false} />`, how would joining work here? Intuitively, you would expect `.a` to override parent `.a`, but compilation to strings wouldn't let us do that. So now we have something that looks like props but doesn't behave like props at all. That seems like a flaw.

Another issue is just the indirection itself. This syntax gives preferential treatment to class names — but it's not clear that in the long term, those would be the preferred way to specify styles anyway. One might want tools to be able to minify them, or isolate them. As you said, it's possible to do this with a build step, but once you have a build step, why write classes at all? For example, Vue prefers the single file component style when you have a build step. If we colocate styles in the same file, it's not clear that we'd even need string names for classes. And then the proposed compilation output doesn't make much sense anymore. Then we might as well start over with something like https://github.com/reactjs/rfcs/pull/159, which already has some comments for future direction.

For the above reasons, I don't think this proposal is viable, so I'm closing it.

---

Your proposal aside, I need to also say a few things about the tone in your comments. Regardless of translation difficulties, statements like "their programming level is too low compared to mine" and "they are still young" are, as @phryneas already noted, very patronizing. While you may disagree with someone else's reasoning, belittling their knowledge or experience level does nothing to advance your arguments, and creates a hostile atmosphere that is not acceptable in this repository.

Moreover, "serious people have gathered here and are solving serious issues, I’ll ask beginners to refrain from cluttering the discussion about adopting a new standard" is the *exact opposite* of the goals of the RFC process — comments from beginners are as appreciated as comments from experts because they help us see issues from a new learner's perspective. That said, again, assuming that everyone who dislikes your idea is a beginner, is also patronizing in its own way.

I hope you can take this feedback to the heart and do better in future discussions.
Cheers, and thank you.

**Comment by MaxmaxmaximusGitHub** [comment]
@gaearon 

> The scope of this RFC is much larger than just React. At the very least, it involves changes to JSX which we discuss here. As @phryneas noted, any changes to JSX are a lot of work for everyone involved to push through because they need to be coordinated with many partner projects, including Babel, ESLint, Prettier, TypeScript, Flow, IDEs and syntax highlighters, and so on. This doesn’t mean it’s insurmountable, but just that the bar for adding something to the JSX grammar is very high.


I retorted this argument above, I said that: AS THE PLAN FOR THE INTRODUCTION OF THE NEW SYNTHAXIS IS HIGH, I NEED TO WRITE PLUGINS NOT MYSELF, AND CALL THE COMMUNITY TO STANDARDIZE. I said that EXACTLY because the inertia of the community is high, I need Facebook support. hence your argument is incorrect.

**Comment by MaxmaxmaximusGitHub** [comment]
plus this syntax allows you to finally do this:


```jsx
function List(props) {
  return <ul .list {...props}>
    
  </ul>
}


function Component() {
  return <List .red></List>
}
```

its render to:
```jsx
<ul className="lisit red></ul>
```

ability at the precompilation stage to competently concatenate dot classes and className is phenomenal =)
---

Adequate guys, do you want me to write a transpiler plugin for this syntax, and add support for it in your IDEs? Who agrees?

**Comment by MaxmaxmaximusGitHub** [comment]
@gaearon 

> If we pass <Foo .a={true} .b={false} />, and then it renders <Bar {...props} .a={false} />, how would joining work here? Intuitively, you would expect .a to override parent .a, but compilation to strings wouldn't let us do that. So now we have something that looks like props but doesn't behave like props at all. That seems like a flaw.

There is no contradiction; the semantics of concatenation do not imply a cascading class system. the concatenation system assumes only syntactic sugar on top of the existing semantics.:

```jsx
function Button (props) {
  return <button .lol={true} {...props}></button>
}

<Button .lol={false}></Button>
```

compile to:
```jsx
function Button (props) {
  return <button {... {
    ...props,
   { (className: true ? "lol" : "") + props.className ? props.className : ""  }
  }}}></button>
}

<Button .lol={false}></Button>
```

**Comment by MaxmaxmaximusGitHub** [comment]
@gaearon if you propose introducing a NEW semantics, which before that was not in react, then you can do it in runtime

you can add object support to new extended props syntax, just add special object `React.props`


```jsx
function Button (props) {
  return <button .lol={true} {...props}></button>
}

<Button .lol={false}></Button>
```

compile to:
```javascript
function Button (props) {
  return React.createElement("button", React.props({classes: {lol: true} }, props) )
}

React.createElement("button", React.props({classes: {lol: false}}) )
```

and that’s it, so you can add different TYPES of properties,

```javascript
React.props({
classes: { class: switcherExp },
key: "key",
ref: React.createRef(),
})
```

this way you can easy create new syntax props, for example `@prop` or` #prop` and give them new semantics

**Comment by MaxmaxmaximusGitHub** [comment]
@gaearon

that way we could create a special runtime string tag for concise compiled code

```jsx
<button .button .__active={user.active}>
  </button>
```

```jsx
React.className = function () {
  // magic
}

function Button(props) {
  return <button
    className={ React.className`button __active${ user.active }` }>
  </button>
}
```

**Comment by MaxmaxmaximusGitHub** [comment]
cheers guys, I came up with a workaround !!!! thank god these guys did not forbid the dollar symbol !!! =)
===

```jsx
<div $chat $active={ false } $opened={ null }></div>
```

I will write a babel plugin that implements the behavior that I described here. the dollar sign will temporarily act as a dot symbol =) this way we can test new syntax

**Comment by MaxmaxmaximusGitHub** [comment]
@gaearon generally we can make a tricky className object like 

{'class1': true, 'class2': false, toString(){ /* magic */ }}

**Comment by MaxmaxmaximusGitHub** [comment]
@gaearon `React.merge` should also be able to combine other React objects.

```javascript
let ref1 = React.createRef()
let ref2 = React.createRef()

let refCompined = React.merge(ref1, ref2)
```

or classLists
```javascript
let classes1 = React.classList({})
let classes2 = React.classList({})

let classesCompined = React.merge(classes1 , classes2 )
```
